### PR TITLE
Parallelize artifact downloads. Update progress bar styling.

### PR DESCRIPTION
**Pull request by IanButterworth**
## This PR
```
% JULIA_DEPOT_PATH=$(mktemp -d): JULIA_PKG_PRECOMPILE_AUTO=0 ./julia --project=../Pkg.jl -t1 -e 'import Pkg; Pkg.activate(temp=true); @time Pkg.add("FFMPEG")'
...
 12.999424 seconds (9.81 M allocations: 915.741 MiB, 2.30% gc time, 73 lock conflicts, 11.69% compilation time: 16% of which was recompilation)
```

## Master
```
% JULIA_DEPOT_PATH=$(mktemp -d): JULIA_PKG_PRECOMPILE_AUTO=0 ./julia-t1 -e 'import Pkg; Pkg.activate(temp=true); @time Pkg.add("FFMPEG")'
...
 23.540455 seconds (9.37 M allocations: 887.289 MiB, 0.82% gc time, 123 lock conflicts, 5.85% compilation time: 18% of which was recompilation)

```

https://github.com/user-attachments/assets/8f1024d6-f255-420c-a70e-becdbfa1dd86

Todo
- [x] parallel progress printing 
- [ ] effective interrupt handling (proven to be hard to do because of Downloads.jl being so async)
  - [ ] needs testing
- [x] error handling & reporting
  - [x] needs testing
- [x] code tidy and remove duplication (will implement a new MiniMultiProgressBar type in a follow-on)
- [x] simple printing for CI etc.
- [x] show progress in bytes 
  - [ ] rate/s ?
  - [ ] total across all artifacts?

### CODE REVIEW

**Comment by IanButterworth** [comment]
Progress printing added 
Screenshot 2024-08-13 at 1 22 18 AM

File: src/Artifacts.jl [hunk 1]
```diff
@@ -394,82 +395,100 @@ function ensure_artifact_installed(name::String, artifacts_toml::String;
...
+            name::String, hash::SHA1, meta::Dict, artifacts_toml::String;
+            platform::AbstractPlatform=HostPlatform(),
+            verbose::Bool=false,
+            quiet_download::Bool=false,
+            io::IO=stderr_f(),
+            progress::Union{Function,Nothing}=nothing)
+
+    errors = Any[]
+    # first try downloading from Pkg server
+    # TODO: only do this if Pkg server knows about this package
```
Comment (KristofferC) [inline src/Artifacts.jl:438]: https://github.com/JuliaLang/Pkg.jl/blob/a7179007182046b28ec9933fa2eaa1c8283a6e00/src/Operations.jl#L918-L931 could be extracted into a function and used here.

File: src/Operations.jl [hunk 2]
```diff
@@ -816,28 +816,132 @@ function collect_artifacts(pkg_root::String; platform::AbstractPlatform=HostPlat
...
     end
     push!(pkg_roots, dirname(env.project_file))
+    used_artifact_tomls = Set{String}()
+    download_jobs = Channel{Function}(Inf)
+
+    print_lock = Base.ReentrantLock() # for non-fancyprint printing
+    download_states = Dict{String, Tuple{Bool,MiniProgressBar}}()
+    errors = Channel{Any}()
+    is_done = false
+    ansi_moveup(n::Int) = string("\e[", n, "A")
```
Comment (KristofferC) [inline src/Operations.jl:840]: Ideally this ANSI stuff could be moved to the progress bar code where there is something like:

```
struct MultiProgressBar
    progress_bars::Vector{MiniProgressBar}
end
```

which takes care of the printing?

**Comment by KristofferC** [comment]
I think the number of significant figures in `Base.format_bytes` is a bit high.

File: src/Artifacts.jl [hunk 3]
```diff
@@ -394,82 +395,100 @@ function ensure_artifact_installed(name::String, artifacts_toml::String;
...
-            if download_success === true
-                return artifact_path(hash)
-            else
-                @debug "Failed to download artifact from Pkg server" download_success
-                push!(errors, (url, download_success))
-            end
+        if isnothing(progress) || verbose == true
+            return try_artifact_download_sources(name, hash, meta, artifacts_toml; platform, verbose, quiet_download, io)
+        else
+            return () -> try_artifact_download_sources(name, hash, meta, artifacts_toml; platform, quiet_download=true, io, progress)
```
Comment (KristofferC) [inline src/Artifacts.jl:421]: This was pretty clever!

**Comment by KristofferC** [comment]
I updated / modernized the progress bar styling a bit. It now looks like:

[Screencast from 08-15-2024 12:19:04 PM.webm](https://github.com/user-attachments/assets/4bc57e42-a9b9-4ce6-bf87-24730ff38242)

**Comment by IanButterworth** [comment]
Error handling is kind of working now, but this log needs proper handling into the formatted message
https://github.com/IanButterworth/Pkg.jl/blob/e8ce1877792f9887d214cd20f0a7dea1f57c12d3/src/PlatformEngines.jl#L646
```
(v1.12) pkg> dev /Users/ian/Documents/GitHub/SimpleLooper.jl
   Resolving package versions...
┌ Error: Hash Mismatch!╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/1
│   Expected sha256:   0000000000000000000000000000000000000000000000000000000000000000
│   Calculated sha256: e65d2f13f2085f2c279830e863292312a72930fee5ba3c792b14c33ce5c5cc58
└ @ Pkg.PlatformEngines ~/Documents/GitHub/Pkg.jl/src/PlatformEngines.jl:646
 Downloading artifacts ╸━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0/1
ERROR: Failed to download some artifacts:

Unable to automatically download/install artifact 'does_not_exist' from sources listed in '/Users/ian/Documents/GitHub/SimpleLooper.jl/Artifacts.toml'.
Sources attempted:
- https://pkg.julialang.org/artifact/0000000000000000000000000000000000000000
    Error: RequestError: HTTP/2 404 while requesting https://pkg.julialang.org/artifact/0000000000000000000000000000000000000000
- https://github.com/staticfloat/small_bin/raw/master/doesnotexist.tar.gz
    Error: RequestError: HTTP/2 404 while requesting https://github.com/staticfloat/small_bin/raw/master/doesnotexist.tar.gz
- https://github.com/staticfloat/small_bin/raw/master/socrates.tar.gz
    Error: Verification failed. Download does not match expected hash


(v1.12) pkg> 
```

**Comment by IanButterworth** [comment]
Screenshot 2024-08-16 at 2 06 58 PM

File: src/Artifacts.jl (hunk 1 above)
Comment (IanButterworth) [reply src/Artifacts.jl:438]: Sounds good, but we'd have to feed the pkg info into this point. I'm unsure of the relationship between packages and artifacts and the pkgserver. Probably best as another PR.

File: src/Operations.jl (hunk 2 above)
Comment (IanButterworth) [reply src/Operations.jl:848]: Yeah. Will handle this in a follow-on. Added a TODO

**Comment by IanButterworth** [comment]
Lets give this a go. @KristofferC and I have both been testing it out.

**Pull request merged**